src/compiler.ts

```diff
@@ -7975,24 +8148,139 @@ export class Compiler extends DiagnosticEmitter {
         prototype.name,
         prototype,
         null,
-        signature,
+        prototype.isAnonymous ? signature.toClosureSignature() : signature,
         contextualTypeArguments
       );
       if (!this.compileFunction(instance)) return this.module.unreachable();
-      this.currentType = contextualSignature.type;
-
     // otherwise compile like a normal function
     } else {
       instance = this.resolver.resolveFunction(prototype, null, contextualTypeArguments);
       if (!instance) return this.module.unreachable();
       this.compileFunction(instance);
-      this.currentType = instance.signature.type;
     }
 
+    // if this anonymous function turns out to be a non-closure, recompile a version
+    // of the function without context, and ignore the other index. Since the index is ignore
+    // the closure version of this function will never be used, and will be optimized out
+    if (instance.closedLocals.size == 0 && instance.prototype.isAnonymous) {
+      this.module.removeFunction(instance.internalName)
+      instance = new Function(
+        instance.prototype.name + "~nonClosure",
```

Comment:
Should be a way to use the name it already had, but was running into some issues. These names aren't exported, so it should be okay

Reply:
That's probably because the old function with the same name is already in `Program#instancesByName`. A better fix would be to not reconstruct the `Function` there, but to modify the existing one if necessary.

Reply:
The challenge with that approach is that all local indexes become invalidated, as they are reduced by 1. Not sure if I'm able to make changes like that after compilation